File: MATLAB code/organise_data.m

clear, clc, close all

folder_path = '..\original_images\';
base_output_path = "..\data\";
ratio = 4;

for resolution_folder = ["FR\", "RR\"]
    
    % Create Empty Resolution folder
    output_path = base_output_path + resolution_folder;
    if exist(output_path, 'dir')
       rmdir(output_path, 's');
    end
    mkdir(output_path);
    
    
    folders = dir(folder_path);
    folders = setdiff({folders.name},{'.','..'})';
    for i = 1:length(folders)
        im_name=folders{i};
        if isfolder(folder_path + "/" + im_name)
            satellite = im_name(1:2);
            switch(satellite)
                case 'W2'
                    sensor = 'WV2';
                case 'W3'
                    sensor = 'WV3';
                case 'W4'
                    sensor = 'WV4';
                case 'GE'
                    sensor = 'GeoEye1';
                otherwise
                    sensor = 'none';
            end
                
            if ~exist(output_path+"\"+satellite, 'dir')
                mkdir(output_path, satellite);
                index=1;
            else
                index = index+1;
            end
    
            % Load Images
            I_PAN = double(imread(folder_path + "/" + im_name+"/RR/PAN.tif"));
            I_MS = double(imread(folder_path + "/" + im_name+"/RR/MS.tif"));
            I_MS_LR = double(imread(folder_path + "/" + im_name+"/RR/MS_LR.tif"));
            I_GT = double(imread(folder_path + "/" + im_name+"/RR/GT.tif"));
            
            % Generate Downgraded Version

            if resolution_folder == "RR/"
                % GT
                I_GT = I_MS_LR;
                
                %   Preparation of image to fuse
                [I_MS_LR, I_PAN] = resize_images(I_MS_LR,I_PAN,ratio,sensor);
                
                % Upsampling
                I_MS = interp23tap(I_MS_LR,ratio);
            end
            % Save Images to new Folder
    
            clear i
            save(output_path+"/"+satellite+"/"+im_name+".mat");
        end
    end
end
disp("Images Generated")

File: dataset_utilities/organise_data.py
import os.path
import shutil

import skimage.io as skio
from scipy.io import savemat

from quality_indexes_toolbox import *

if __name__ == '__main__':
    folder_path = '..\\original_images\\'
    folder_path = "C:\\Users\\pmans\\Documents\\Magistrale\\Remote Sensing\\Progetto\\PAirMax\\"
    base_output_path = "..\\data\\"
    ratio = 4
    index = -1

    if os.path.exists(base_output_path):
        shutil.rmtree(base_output_path)
    os.mkdir(base_output_path)

    for resolution in ["FR", "RR"]:
        # Create Output Folder
        output_path = f"{base_output_path}\\{resolution}"
        if os.path.exists(output_path):
            shutil.rmtree(output_path)
        os.mkdir(output_path)

        # Loop Over Images of the Dataset
        for img in os.listdir(folder_path):
            spl_str = img.split("_")
            if len(spl_str) != 3:
                continue

            # Extract Sensor Name
            sensor_img = spl_str[0]

            if sensor_img == "W2":
                sensor = "WV2"
            elif sensor_img == "W3":
                sensor = "WV3"
            elif sensor_img == "W4":
                sensor = "WV4"
            elif sensor_img == "GE":
                sensor = "GeoEye1"
            else:
                sensor = "none"

            # Create satellite output folder
            satellite_folder = f"{output_path}\\{sensor_img}"
            if not os.path.exists(satellite_folder):
                os.mkdir(satellite_folder)
                index = 1
            else:
                index += 1

            # Read Full Resolution Images (According to Reduced Resolution Protocol)
            I_PAN = skio.imread(f"{folder_path}\\{img}\\RR/PAN.tif").astype("double")
            I_MS = skio.imread(f"{folder_path}\\{img}\\RR/MS.tif").astype("double")
            I_MS_LR = skio.imread(f"{folder_path}\\{img}\\RR/MS_LR.tif").astype("double")
            I_GT = skio.imread(f"{folder_path}\\{img}\\RR/GT.tif").astype("double")

            # Generate Downgraded Version
            if resolution == "RR":
                I_GT = I_MS_LR
                [I_MS_LR, I_PAN] = resize_images(I_MS_LR, I_PAN, ratio, sensor)
                I_MS = interp23(I_MS_LR, ratio)

            savemat(f"{satellite_folder}\\{img}.mat", dict(I_PAN=I_PAN, I_GT=I_GT, I_MS=I_MS, I_MS_LR=I_MS_LR))

    print("Images Organization Completed")
